app/utils/database.py

```diff
@@ -158,6 +171,50 @@ class Herd(BaseModel):
     longitude = FloatField(null=True)
     coordinates_privacy = CharField(15, null=True)
 
+    def filtered_dict(self, user=None):
+        """
+        Returns the model data filtered by the access level of the given `user`.
+        """
+
+        access_level = 'public'
+        if user and user.is_admin:
+            access_level = 'private'
+        elif user:
+            for role in user.privileges:
+                if role['level'] in ['specialist', 'manager']:
+                    if role['genebank'] == self.genebank.id:
+                        access_level = 'private'
+                        break
+                elif role['level'] == 'owner':
+                    if role['herd'] == self.id:
+                        access_level = 'private'
+                        break
+                    access_level = 'authenticated'
+
+        data = self.as_dict()
+
+        # prune system data
+        del data['email_verified']
+
+        levels = ['public', 'authenticated', 'private']
+        # prune data according to access level
+        for field in [f for f in data.keys() if f.endswith('_privacy')]:
+            # remove values if access_level is less than required
+            field_level = data[field] or 'private'
+            if levels.index(access_level) < levels.index(field_level):
+                if field == 'coordinates_privacy':
+                    del data['latitude']
+                    del data['longitude']
+                else:
+                    target_field = field[:-8]
```

Comment:
More apparent what the `-8` does by explicitly writing the expression it came from: 
```python
target_field = field[:-len('_privacy')]
```